Retry flows API

hochgi opened the pull request:
PR for https://github.com/akka/akka-stream-contrib/issues/25

hochgi commented:
Is the `concat` APIs possible? If we have the retries cycling in a loop, adding more elements in a closed loop would require unbounded buffering or it would deadlock.

Maybe it can be done, but not with the suggested API. We must not add elements to the loop unless the same amount of elements is removed from the loop.

So just playing around, what if we add a special type:

```scala
sealed trait RetryFork[T]
case object Emit extends RetryFork[Nothing]
case class Retry[T](head: T, tail: RetryFork[T]) extends RetryFork[T]
case class EmitAndRetry[T](head: T, tail: RetryFork[T]) extends RetryFork[T]
```

And have the state `S` be of this type, resulting in an API similar to:

```scala
def concat[I,O,S,M](flow: Flow[(I,S),(Try[O],S),M])(retryWith: S => RetryFork[(I,S)])
```
 Thus enabling to generate more elements to drop into the retry loop, but doing so lazily, so new elements are only generated when a previous element left the cycle

Not sure all of this makes sense, and probably an overkill for a generic API, but just thinking out loud here...

viktorklang commented:
There is no generic way to know if an element leaves the loop.

On Wed, Jun 22, 2016 at 3:51 AM, Gilad Hoch <notifications@github.com>
wrote:

> Is the concat APIs possible? If we have the retries cycling in a loop,
> adding more elements in a closed loop would require unbounded buffering or
> it would deadlock.
>
> Maybe it can be done, but not with the suggested API. We must not add
> elements to the loop unless the same amount of elements is removed from the
> loop.
>
> So just playing around, what if we add a special type:
>
> sealed trait RetryFork[T]case object Emit extends RetryFork[Nothing]case class Retry[T](head: T, tail: RetryFork[T]) extends RetryFork[T]case class EmitAndRetry[T](head: T, tail: RetryFork[T]) extends RetryFork[T]
>
> And have the state S be of this type, resulting in an API similar to:
>
> def concat[I,O,S,M](flow: Flow[(I,S),(Try[O],S),M])(retryWith: S => RetryFork[(I,S)])
>
> Thus enabling to generate more elements to drop into the retry loop, but
> doing so lazily, so new elements are only generated when a previous element
> left the cycle
>
> Not sure all of this makes sense, and probably an overkill for a generic
> API, but just thinking out loud here...
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/akka/akka-stream-contrib/pull/26#issuecomment-227622133>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAqd1jyP6JvtiT-rKZEbooMTNpR-8BTks5qOJU_gaJpZM4I5cEY>
> .
>



-- 
Cheers,
√

hochgi commented:
@viktorklang not for a specific element, but we can wrap the entire graph wih stages that keep track of how many element are in the cycle. e.g:
![retry graph](https://gist.githubusercontent.com/hochgi/c8ee3273034fff8958d306459072fb22/raw/764cb5a5e4d02db1284ee51fe3feb69a119fc976/retry-stream.png)

viktorklang commented:
@hochgi No, I meant, for a Graph X, there is no generic way to track element "loss" :)

hochgi commented:
you mean inside the given flow? a flow isn't limited to 1:1 input-output, but that will be of course documented as a caveat users should be aware to.

What I have in mind, is the obvious usecase (that got me into this in the first place) of a connection pool flow.
something like `Http().superPool()`, and I assume this flow is guaranteed to be 1:1 input-output.

viktorklang commented:
@hochgi Yes, the reason I say this is because if all a user has is a Flow[I,O,M] it is impossible to know if it is 1:1 or not.

hochgi commented:
@viktorklang OK, I'm fine with documenting this as a warning. it's something the user can, but should not do. like passing `null`s...

viktorklang commented:
@hochgi passing nulls as elements is not allowed :)

hochgi commented:
current status: tried implementing the basic functionality described in the image
![retry graph](https://gist.githubusercontent.com/hochgi/c8ee3273034fff8958d306459072fb22/raw/764cb5a5e4d02db1284ee51fe3feb69a119fc976/retry-stream.png)
but stream won't shut down. implementation:
https://github.com/hochgi/akka-stream-contrib/commit/9812757ff1103eaa4335a4c10212e2ad531672a1
not sure what I missed, will dig into it later.

hochgi commented:
The simple retry works now. currently thinking how to implement the `retryConcat`, and if the API is reasonable.

hochgi commented:
in the last amended commit I removed the async APIs, and implemented the concat according to this diagram:
![retry concat graph](https://gist.githubusercontent.com/hochgi/c8ee3273034fff8958d306459072fb22/raw/c1e8cf5991e7156a372a027c09deb079fd50609f/retry-concat-stream-diagram.png)
there is already test and some very minimal documentation.
To get this PR ready for merge I want to add more docs.
Especially, I want to document the caveats discussed here (e.g: Must provide 1:1 flows)
other than that, I think the code is ready for reviews.
cc @viktorklang

hochgi commented:
Documentation is written, and commits are squashed.

patriknw commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,245 @@
+          elementsInCycleCount += 1
+        }
+
+        override def onUpstreamFinish() = {
```
is there a need for `setKeepGoing(true)`?

hochgi replied:
I don't think so. in case there are no elements in the cycle, the stage should complete regularly.

patriknw commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,245 @@
+    override def createLogic(attributes: Attributes) = new GraphStageLogic(shape) {
+
+      import shape._
+      var elementsInCycleCount = 0
```
`Long` to avoid (theoretical) `Int` wrap around issues?

hochgi replied:
OK, it's not likely, but is possible that the given flow will aggregate more than `Int.MaxValue` elements. I will change to `Long`

hochgi commented:
@patriknw changed `Int` to `Long` and amended the commit :)

drewhk commented:
You can wrap the Flow with a BidiFlow that tracks that not more elements are emitted than have been consumed, otherwise fails the stream.

drewhk commented:
nice diagrams btw.

drewhk commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,245 @@
+
+  /**
+   * Factory for a retry flow. Given a flow which recieves a tuple of "input" & "state",
+   * and emits a tuples of a Try of "output" & "state", and given a function that given a state
```
very, very long sentence :) Please try to refactor.

drewhk commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,245 @@
+
+  private[akka] class CountElementsAndInterceptCompletion[T] extends GraphStage[FanInShape2[T, Long, T]] {
+    override val shape = new FanInShape2[T, Long, T]("countElementsAndInterceptCompletion")
+    override def createLogic(attributes: Attributes) = new GraphStageLogic(shape) {
```
extend OutHandler to reduce allocations

hochgi replied:
Not sure what you mean... extends where?

drewhk replied:
I meant `new GraphStageLogic(shape) with OutHandler` and then `setHandler(out, this)`.

drewhk commented:
One option for improvement is to make the increment/decrement stages one BidiStage instead, then there is no cost of communication (and possible races) between the two sides.

drewhk commented on src/test/scala/akka/stream/contrib/RetrySpec.scala:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  "Retry" should {
+    "retry ints according to their parity" in {
```
more tests are needed, especially around completion, and also testing deadlock-freedom.

hochgi replied:
indeed. will do

drewhk commented:
I am not sure that this graph is necessary. I think it might be far simpler to concentrate all the logic and bookkeeping to one BidiStage. It would look like this:

```
       +-------------------+
 +-----> Flow              +----+
 |     +-------------------+    |
 |                              |
 |     +-------------------+    |
 +-----+ RertryCoordinator +----+
       |                   |
 +----->                   +---->
       +-------------------+
```

hochgi commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,245 @@
+   * break down a "heavy" element which failed into multiple "thin" elements, that may succeed
+   * indevidually. Since it's very easy to inflate elements in retry cycle, there's also a
+   * limit parameter, that will limit the amount of generated elements by the `retryWith` function,
+   * and will fail the stage if that limit is exceeded.
```
Maybe the API can be expanded to receive an overflow strategy, like the buffer stage.
I have thought of that, and decided it's an overkill for now.  But since it's the contrib, I assume that if down the road we break API and add another overflow strategy parameter, it'll be OK.

anyway, would you want me to add that now before accepting this PR?

P.S. as a default behaviour, IMO, failing the stage is better. as a user of this library, I would want to know in the development phase I did something wrong, instead of discovering in production I'm losing data silently.

drewhk replied:
ok, agreed!

hochgi commented:
@drewhk the BidiFlow idea is great! wish I'd thought of it earlier...
regarding failing the stage if the given flow output more elements than it should, I think it's not a good idea.
you can only detect if the are too many elements coming out, but what about a flow that reduces the number of elements? If we can't guarantee this safely, as @viktorklang pointed out, I think we shouldn't do it at all, and leave this as the user responsibility.

drewhk commented:
> regarding failing the stage if the given flow output more elements than it should, I think it's not a good idea.
you can only detect if the are too many elements coming out, but what about a flow that reduces the number of elements? If we can't guarantee this safely, as @viktorklang pointed out, I think we shouldn't do it at all, and leave this as the user responsibility.

Ok, you convinced me.

hochgi commented:
@drewhk I have implemented from scratch using a bidi flow. that was a great idea! thanks :)
So now, I just need to write some additional tests & improve documentation.

drewhk commented:
Ok, now I look at this finally :)

drewhk commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  private[akka] class RetryCoordinator[I, S, O](retryWith: S => Option[(I, S)]) extends GraphStage[BidiShape[(I, S), (Try[O], S), (Try[O], S), (I, S)]] {
+    val in1 = Inlet[(I, S)]("RertryCoordinator.in")
```
I recommend using more descriptive names for these ports to so the code reads nicer.

hochgi replied:
Any suggestions?

drewhk commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,244 @@
+
+      setHandler(out1, new OutHandler {
+        override def onPull() =
+          pull(in1)
```
it would be good to write up the invariants in a comment. Like the above line relies on the invariant that if `out1` is pulled then `in1` is available for pulling.

drewhk replied:
also, it would be probably helpful to add a comment describing a duty-cycle of the stage.

drewhk commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,244 @@
+      setHandler(in1, new InHandler {
+        override def onPush() = {
+          val is = grab(in1)
+          if (!hasBeenPulled(in2)) pull(in2)
```
hm, does not make sense instead to pull through the wrapped stage? I.e. pulling of out1 results this stage pulling the wrapped flow, which in turn will pull out2, which in turn will pull in1? That is a control flow that might be more understandable.

drewhk replied:
to describe what I mean, the duty cycle with one retry would look like this
 - out1 is pulled by external world, we pull in2 (the wrapped flow)
 - out2 is pulled by wrapped Flow, we pull in1 (the external world)
 - in1 is pushed by external world, we feed result in out2 (to wrapped flow)
 - wrapped flow pushes failure through in2. We decide to retry (instead of pushing to external world), and pull in2 (the wrapped flow)
 - out2 is pulled by wrapped Flow, we provide retry (instead of pulling external world) through out2
 - wrapped flow pushes success through in2, we push it through out1 to external world
 - cycle starts from step 1 again

I quite possibly miss some details here that you already found out, so correct me on any point here.

hochgi replied:
That was my initial logic. But the wrapped flow may pull eagerly without being pulled, always storing one element ready for downstream immediately.

drewhk replied:
Aargh, yes the flow can pre-pull, I forgot that :)

hochgi replied:
well, I think I managed to pull off with your suggested duty cycle (at least on the simpler case, and check if the flow pre-pulls elements)

drewhk replied:
hm, one difference is that on out1.onPull you immediately pull in1, and not the internal Flow.

hochgi replied:
not exactly. in the revised code we have:
```scala
setHandler(out1, new OutHandler {
  override def onPull() = {
    if (isAvailable(out2)) pull(in1)
    else pull(in2)
  }
})
```

So we only pull in1 if the internal flow has already pulled, which means I won't get out2.onPull.
in case out2 isn't available, we do pull in2 and not in2.

drewhk replied:
Oh, why do I don't see the right version here...

drewhk replied:
now I do, weird.

hochgi replied:
yeah... github is being weird lately. I needed to refresh the page a couple of times before getting the new code I committed yesterday :/

drewhk commented on src/test/scala/akka/stream/contrib/RetrySpec.scala:
```diff
@@ -0,0 +1,264 @@
+      killSwitch.abort(failedElem.failed.get)
+      sink.request(1)
+      sink.expectError(failedElem.failed.get)
+    }
```
I think it would be important to have a version where the flow passed to Retry itself is terminated via a KillSwitch

hochgi replied:
done :)

drewhk replied:
ok, I am blind. I was looking for something like `Retry(innerFlow.via(KillSwitch))` but `innerFlow` already had it applied, sorry for the noise.

hochgi replied:
No you arent blind. These are new tests :-)

drewhk replied:
ok great, doesn't matter in the end :D

drewhk commented:
The logic looks correct to me, so if a few tests are added where the inner flow is killed I would be happy to merge this :)

hochgi commented:
There are already 6 such tests. 3 for retry and same 3 for retry.concat
What other scenarios are missing?

drewhk commented:
Nothing is missing, I was looking at the wrong place (this time not GHs fault). LGTM

drewhk commented:
Can you squash the commits into one? Then I would like to merge this (unless other volunteers decide to review).

hochgi commented:
@drewhk commits are squashed!

johanandren commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,251 @@
+    val in1 = Inlet[(I, S)]("Rertry.ext.in")
+    val out1 = Outlet[(Try[O], S)]("Rertry.ext.out")
+    val in2 = Inlet[(Try[O], S)]("Rertry.int.in")
+    val out2 = Outlet[(I, S)]("Rertry.int.out")
```
Rertry -> Retry in all these

hochgi replied:
@johanandren oops... typo. fixed now :-)

johanandren commented:
LGTM

ktoso commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,251 @@
+   *
+   * IMPORTANT CAVEAT:
+   * The given flow must not change the number of elements passing through it (i.e. it should output
+   * exactly one element for every received element). Ignoring this, will have an unpredicted result.
```
And may result in a deadlock

ktoso commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,251 @@
+  /**
+   * Retry flow factory. given a flow that produces `Try`s, this wrapping flow may be used to try
+   * and pass failed elements through the flow again. More accurately, the given flow consumes a tuple
+   * of "input" & "state", and produces a tuple of `Try` of "output" and "state.
```
```
`input`

`state`
```

ktoso commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,251 @@
+   * @tparam O output elements type
+   * @tparam S state to create a new (input,state): (I,S) to retry with
+   * @tparam M materialized value type
+   * @return `Graph[FlowShape[(I,S),Try[O]],M]` instance
```
remove `@return` doc, it's not helping – we have return types for that ;)

ktoso commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,251 @@
+   *        and retry with a new pair of input & new state we get from this function.
+   * @tparam I input elements type
+   * @tparam O output elements type
+   * @tparam S state to create a new (input,state): (I,S) to retry with
```
```
@tparam S state to create a new `(I, S)` to retry with
`

ktoso commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,251 @@
+
+  /**
+   * Factory for multiple retries flow. similar to the simple retry, but this will allow to
+   * break down a "heavy" element which failed into multiple "thin" elements, that may succeed indevidually.
```
individually

ktoso commented on src/main/scala/akka/stream/contrib/Retry.scala:
```diff
@@ -0,0 +1,251 @@
+  /**
+   * Factory for multiple retries flow. similar to the simple retry, but this will allow to
+   * break down a "heavy" element which failed into multiple "thin" elements, that may succeed indevidually.
+   * Since it's very easy to inflate elements in retry cycle, there's also a limit parameter,
```
very easy => easy, we tend to avoid "very" as it depends on who's reading this

ktoso commented:
LGTM, would be nice to have the graph / docs persited somewhere.
I guess we should think about docs for this project too?

hochgi commented:
@ktoso I agree. docs for contrib would be great. not sure where it should be though...
streams cookbook?
here in GH as wiki pages?
maybe the new SO documentation would be a good place to start?

ktoso commented:
> @ktoso I agree. docs for contrib would be great. not sure where it should be though...

No, proper docs – we need docs for Reactive Kafka anyway, so we'll figure out a common way to do it.

> streams cookbook?

No, since it's in the wrong place then

> here in GH as wiki pages?

not wiki pages, docs should be compiled and tested. same as in the akka project.

> maybe the new SO documentation would be a good place to start?

same point as above, reference docs can be compiled and tested + comitted together with initial feature PR. SO is a fun thing, but it's additional.

ktoso commented:
Anyhow, this feature is a nice one! Seems it was reviewed by 3 already even :-)
